packages/in_app_purchase/in_app_purchase_storekit/darwin/Classes/InAppPurchasePlugin.swift

```diff
@@ -253,16 +253,24 @@ public class InAppPurchasePlugin: NSObject, FlutterPlugin, FIAInAppPurchaseAPI {
     let pendingTransactions = getPaymentQueueHandler().getUnfinishedTransactions()
 
     for transaction in pendingTransactions {
+      // finishTransaction() cannot be called on a Transaction with a current purchasing state
+      // https://developer.apple.com/documentation/storekit/skpaymentqueue/1506003-finishtransaction
+      guard transaction.transactionState != SKPaymentTransactionState.purchasing else {
+        continue
+      }
+
       // If the user cancels the purchase dialog we won't have a transactionIdentifier.
       // So if it is null AND a transaction in the pendingTransactions list has
       // also a null transactionIdentifier we check for equal product identifiers.
-      if transaction.transactionIdentifier == transactionIdentifier
-        || (transactionIdentifier == nil
-          && transaction.transactionIdentifier == nil
-          && transaction.payment.productIdentifier == productIdentifier)
-      {
-        getPaymentQueueHandler().finish(transaction)
+      let matchesTransactionIdentifier = transaction.transactionIdentifier == transactionIdentifier
+      let isRestoringTransaction =
```

Comment:
Can you comment why this is "restoring transaction"? The comment above mentions "canceling", but not "restoring"

Reply:
oops.

Reply:
I ended up trying to look up where this behaviour is documented to signify cancellation, since I copied the comment from the objc version. Looks like we might be able to check for SKErrorPaymentCancelled instead, so I left a TODO as well, cuz its not related to this PR.